src/main/java/com/direwolf20/buildinggadgets/common/building/implementation/IBuildingMode.java

```diff
@@ -0,0 +1,59 @@
+package com.direwolf20.buildinggadgets.common.building.implementation;
+
+import com.direwolf20.buildinggadgets.common.building.placement.Context;
+import com.direwolf20.buildinggadgets.common.building.placement.IBlockProvider;
+import com.direwolf20.buildinggadgets.common.building.placement.IPlacementSequence;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.client.resources.I18n;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import java.util.function.BiPredicate;
+
+/**
+ * Represents a mode that can be used by some gadget,
+ */
+public interface IBuildingMode {
+
+    /**
+     * Iterator that supplies raw coordinates that hasn't been filtered yet.
+     */
+    IPlacementSequence computeCoordinates(EntityPlayer player, BlockPos hit, EnumFacing sideHit, ItemStack tool);
+
+    IBlockProvider getBlockProvider(ItemStack tool);
```

Comment:
This will sadly force us to only support known implementations... So maybe create an BlockProvider Capability, which could be returned here?

Reply:
Good idea